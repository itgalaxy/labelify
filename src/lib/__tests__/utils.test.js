"use strict";

const os = require("os");
const path = require("path");
const nock = require("nock");
const utils = require("../utils");
const pkg = require("../../../package");

const {
  buildGotOptions,
  hexRegExp,
  request,
  getLinks,
  getPlatformAndEndpoint,
  getModulePath,
  getToken,
  linkWalker,
  getProjectMeta,
  resolveRepositoryURL,
  loadConfig
} = utils;

const fixturesDir = path.resolve(__dirname, "fixtures");
const configsDir = path.join(fixturesDir, "configs");

describe("utils", () => {
  it("should success execute request", () => {
    expect.assertions(1);

    nock("https://labelify.com")
      .get("/")
      .reply(200, "Hello world!");

    return request("https://labelify.com", {}).then(result => {
      expect(result.body).toBe("Hello world!");

      return result;
    });
  });

  it("should success build got options", () => {
    expect.assertions(5);

    expect(buildGotOptions()).toEqual({
      headers: {
        "user-agent": `labelify/${pkg.version} (https://github.com/itgalaxy/labelify)`
      }
    });

    expect(buildGotOptions({})).toEqual({
      headers: {
        "user-agent": `labelify/${pkg.version} (https://github.com/itgalaxy/labelify)`
      }
    });

    expect(buildGotOptions({ foo: "bar" })).toEqual({
      foo: "bar",
      headers: {
        "user-agent": `labelify/${pkg.version} (https://github.com/itgalaxy/labelify)`
      }
    });

    expect(buildGotOptions({ headers: {} })).toEqual({
      headers: {
        "user-agent": `labelify/${pkg.version} (https://github.com/itgalaxy/labelify)`
      }
    });

    expect(buildGotOptions({ headers: { foo: "bar" } })).toEqual({
      headers: {
        foo: "bar",
        "user-agent": `labelify/${pkg.version} (https://github.com/itgalaxy/labelify)`
      }
    });
  });

  it("should success get links", () => {
    expect.assertions(2);

    expect(getLinks("")).toMatchSnapshot();
    expect(
      getLinks(
        '<https://api.github.com/repositories/131302483/labels?page=2>; rel="next", ' +
          '<https://api.github.com/repositories/131302483/labels?page=2>; rel="last"\n' +
          '<https://api.github.com/repositories/131302483/labels?page=1>; rel="prev", ' +
          '<https://api.github.com/repositories/131302483/labels?page=1>; rel="first"'
      )
    ).toMatchSnapshot();
  });

  it("should success get module path", () => {
    expect.assertions(3);

    expect(
      path.relative(process.cwd(), getModulePath("not-exists", "resolve-from"))
    ).toMatchSnapshot();
    expect(
      path.relative(
        process.cwd(),
        getModulePath("node_modules", "resolve-from")
      )
    ).toMatchSnapshot();
    expect(() => getModulePath("foo", "bar")).toThrowErrorMatchingSnapshot();
  });

  it("should success get platform and endpoint for github", () =>
    expect(
      getPlatformAndEndpoint("https://github.com/itgalaxy/labelify.git")
    ).toMatchSnapshot());

  it("should success get platform and endpoint for github (git)", () =>
    expect(
      getPlatformAndEndpoint("git+https://github.com/111/222.git")
    ).toMatchSnapshot());

  it("should success get platform and endpoint for github (git shortcut)", () =>
    expect(getPlatformAndEndpoint("github:111/222")).toMatchSnapshot());

  it("should success get platform and endpoint for gitlab", () =>
    expect(
      getPlatformAndEndpoint("https://gitlab.com/itgalaxy/labelify.git")
    ).toMatchSnapshot());

  it("should success get platform and endpoint for gitlab (git shortcut)", () =>
    expect(getPlatformAndEndpoint("gitlab:111/222")).toMatchSnapshot());

  it("should success get platform and endpoint for github (self-hosted)", () =>
    expect(
      getPlatformAndEndpoint(
        "https://github.yourcompany.com/itgalaxy/labelify.git",
        {
          platform: "github"
        }
      )
    ).toMatchSnapshot());

  it("should success get platform and endpoint for github (self-hosted with nonstandard protocol)", () =>
    expect(
      getPlatformAndEndpoint("ssh://git@mydomain.com/abc/def", {
        platform: "github"
      })
    ).toMatchSnapshot());

  it("should success get platform and endpoint for gitlab (self-hosted)", () =>
    expect(
      getPlatformAndEndpoint(
        "https://gitlab.company.com/itgalaxy/labelify.git",
        {
          platform: "gitlab"
        }
      )
    ).toMatchSnapshot());

  it("should success get platform and endpoint for gitlab (self-hosted with nonstandard protocol)", () =>
    expect(
      getPlatformAndEndpoint("ssh://git@mydomain.com/abc/def", {
        platform: "gitlab"
      })
    ).toMatchSnapshot());

  it("should success get platform and endpoint for custom platform and custom endpoint", () =>
    expect(
      getPlatformAndEndpoint(
        "https://my.custom.platform.com/itgalaxy/labelify.git",
        {
          endpoint: "https://my.custom.platform.com/custom/api/url",
          platform: "github"
        }
      )
    ).toMatchSnapshot());

  it("should success get platform and endpoint for custom platform and custom endpoint (2)", () =>
    expect(
      getPlatformAndEndpoint(
        "https://my.custom.platform.com/itgalaxy/labelify.git",
        {
          endpoint: "https://my.custom.platform.com/custom/api/url",
          platform: "gitlab"
        }
      )
    ).toMatchSnapshot());

  it("should success get platform and endpoint (shortcut syntax)", () =>
    expect(
      getPlatformAndEndpoint("github:itgalaxy/labelify")
    ).toMatchSnapshot());

  it("should throw error when repository URL doesn't contains user", () =>
    expect(() =>
      getPlatformAndEndpoint("https://gitlab.company.com", {
        platform: "gitlab"
      })
    ).toThrowErrorMatchingSnapshot());

  it("should throw error when repository URL doesn't contains project", () =>
    expect(() =>
      getPlatformAndEndpoint("https://gitlab.company.com/user", {
        platform: "gitlab"
      })
    ).toThrowErrorMatchingSnapshot());

  it("should throw error when repository URL is not passed", () =>
    expect(() => getPlatformAndEndpoint()).toThrowErrorMatchingSnapshot());

  it("should throw error when repository URL is not supported", () =>
    expect(() =>
      getPlatformAndEndpoint("https://v8.googlecode.com/svn/trunk/")
    ).toThrowErrorMatchingSnapshot());

  it("should throw error when repository URL is not supported (2)", () =>
    expect(() =>
      getPlatformAndEndpoint("bitbucket:user/repo")
    ).toThrowErrorMatchingSnapshot());

  it("should throw error when repository URL is self hosted platform without 'platform' option", () =>
    expect(() =>
      getPlatformAndEndpoint(
        "https://github.yourcompany.com/itgalaxy/labelify.git"
      )
    ).toThrowErrorMatchingSnapshot());

  it("should success resolve repository URL in package.json with string repository field", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();

    process.chdir(
      path.join(fixturesDir, "package-json-with-string-repository-field")
    );

    return expect(
      resolveRepositoryURL().then(repositoryURL => {
        process.chdir(oldProcessCwd);

        return repositoryURL;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success resolve repository URL in package.json with object repository field", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();

    process.chdir(
      path.join(fixturesDir, "package-json-with-object-repository-field")
    );

    return expect(
      resolveRepositoryURL().then(repositoryURL => {
        process.chdir(oldProcessCwd);

        return repositoryURL;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should throw the error when `package.json` don't have repository field", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();

    process.chdir(
      path.join(fixturesDir, "package-json-without-repository-field")
    );

    return expect(
      resolveRepositoryURL().catch(error => {
        process.chdir(oldProcessCwd);

        throw error;
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should throw the error when `package.json` contain broken repository object field", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();

    process.chdir(
      path.join(fixturesDir, "package-json-with-broken-object-repository-field")
    );

    return expect(
      resolveRepositoryURL().catch(error => {
        process.chdir(oldProcessCwd);

        throw error;
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should throw error when `package.json` doesn't found", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();

    process.chdir("/tmp");

    return expect(
      resolveRepositoryURL().catch(error => {
        process.chdir(oldProcessCwd);

        throw error;
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should success get project meta", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();

    process.chdir(path.join(fixturesDir));

    return expect(
      getProjectMeta().then(projectMeta => {
        process.chdir(oldProcessCwd);

        return projectMeta;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success get project meta (2)", () => {
    expect.assertions(1);

    return expect(
      getProjectMeta({
        endpoint: "https://api.github.com/repos/itgalaxy/labelify/labels",
        platform: "github"
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success get token", () => {
    expect.assertions(4);

    expect(typeof getToken()).toBe("undefined");

    process.env.TOKEN = "TOKEN";

    expect(getToken()).toBe(process.env.TOKEN);

    delete process.env.TOKEN;

    process.env.GITHUB_TOKEN = "GITHUB_TOKEN";

    expect(getToken()).toBe(process.env.GITHUB_TOKEN);

    delete process.env.GITHUB_TOKEN;

    process.env.GITLAB_TOKEN = "GITLAB_TOKEN";

    expect(getToken()).toBe(process.env.GITLAB_TOKEN);

    delete process.env.GITLAB_TOKEN;
  });

  it("should success execute hex regexp", () => {
    expect.assertions(7);

    expect(hexRegExp.test("#fff")).toBe(true);
    expect(hexRegExp.test("#ffffff")).toBe(true);
    expect(hexRegExp.test("#FFF")).toBe(true);
    expect(hexRegExp.test("#FFFFFF")).toBe(true);
    expect(hexRegExp.test("FFFFFF")).toBe(false);
    expect(hexRegExp.test("#f")).toBe(false);
    expect(hexRegExp.test("#fffffff")).toBe(false);
  });

  it("should success walk on the links", () => {
    expect.assertions(1);

    const url = "https://labelify.com";

    nock(url)
      .get("/")
      .reply(200, "Page 1!", {
        link: `<${url}?page=2>; rel="next",
        <${url}?page=2>; rel="last"
        <${url}?page=1>; rel="prev",
        <${url}?page=1>; rel="first"`
      });

    nock(url)
      .get("/?page=2")
      .reply(200, "Page 2!", {
        link: `<${url}?page=2>; rel="last"
        <${url}?page=1>; rel="prev",
        <${url}?page=1>; rel="first"`
      });

    return expect(linkWalker(url)).resolves.toMatchSnapshot();
  });

  it("should success build config", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "simple");

    process.chdir(configDir);

    return expect(
      loadConfig().then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should return null when can't load config", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();

    process.chdir(os.tmpdir());

    return expect(
      loadConfig().then(config => {
        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build config using 'configFile' option", () => {
    expect.assertions(1);

    const configDir = path.join(configsDir, "simple");

    return expect(
      loadConfig({
        configFile: path.join(configDir, "labelify.config.js")
      }).then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build config with package 'extends'", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "package-extends");

    process.chdir(configDir);

    return expect(
      loadConfig().then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build config with package 'extends' and 'configFile' option", () => {
    expect.assertions(1);

    const configDir = path.join(configsDir, "package-extends");

    return expect(
      loadConfig({
        configFile: path.join(configDir, "labelify.config.js")
      }).then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build with relative path 'extends'", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "relative-path-extends");

    process.chdir(configDir);

    return expect(
      loadConfig().then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build with absolute path 'extends'", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "absolute-path-extends");

    process.chdir(configDir);

    return expect(
      loadConfig().then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build with relative path 'extends' and 'configFile' option", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "relative-path-extends");

    process.chdir(configDir);

    return expect(
      loadConfig({
        configFile: path.join(configDir, "labelify.config.js")
      }).then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build with absolute path 'extends' and 'configFile' option", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "absolute-path-extends");

    process.chdir(configDir);

    return expect(
      loadConfig({
        configFile: path.join(configDir, "labelify.config.js")
      }).then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build config with array of 'extends'", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "array-extends");

    process.chdir(configDir);

    return expect(
      loadConfig().then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build config using 'configBase' option", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "config-base");

    process.chdir(configDir);

    return expect(
      loadConfig({
        configBasedir: path.join(configDir, "custom-directory")
      }).then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should success build config using 'configFile' and 'configBase' options", () => {
    expect.assertions(1);

    const oldProcessCwd = process.cwd();
    const configDir = path.join(configsDir, "config-base");

    process.chdir(configDir);

    return expect(
      loadConfig({
        configBasedir: path.join(configDir, "custom-directory"),
        configFile: path.join(configDir, "labelify.config.js")
      }).then(config => {
        config.filepath = path.relative(process.cwd(), config.filepath);

        process.chdir(oldProcessCwd);

        return config;
      })
    ).resolves.toMatchSnapshot();
  });

  it("should throw error when config not found using `configFile` option", () => {
    expect.assertions(1);

    return expect(
      loadConfig({
        configFile: "not-found"
      }).catch(error => {
        error.message = error.message.replace(`${process.cwd()}/`, "");

        throw error;
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should throw error when extended config not found using `configFile` option", () => {
    expect.assertions(1);

    const configDir = path.join(configsDir, "config-base");

    return expect(
      loadConfig({
        configFile: path.join(configDir, "labelify.config.js")
      }).catch(error => {
        error.message = error.message.replace(`${process.cwd()}/`, "");

        throw error;
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
