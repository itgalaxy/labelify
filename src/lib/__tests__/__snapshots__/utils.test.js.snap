// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils should return null when can't load config 1`] = `null`;

exports[`utils should success build config 1`] = `
Object {
  "config": Object {
    "config": Object {
      "endpoint": "http://example.com/foo/bar",
      "platform": "github",
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#111111",
        "name": "1",
      },
      Object {
        "color": "#222222",
        "name": "2",
      },
      Object {
        "color": "#333333",
        "name": "3",
      },
      Object {
        "color": "#444444",
        "name": "4",
      },
      Object {
        "color": "#555555",
        "name": "5",
      },
      Object {
        "color": "#666666",
        "name": "6",
      },
      Object {
        "color": "#777777",
        "name": "7",
      },
      Object {
        "color": "#888888",
        "name": "8",
      },
      Object {
        "color": "#999999",
        "name": "9",
      },
      Object {
        "color": "#aaaaaa",
        "name": "a",
      },
      Object {
        "color": "#bbbbbb",
        "name": "b",
      },
      Object {
        "color": "#cccccc",
        "name": "c",
      },
      Object {
        "color": "#dddddd",
        "name": "d",
      },
      Object {
        "color": "#eeeeee",
        "name": "e",
      },
      Object {
        "color": "#ffffff",
        "name": "f",
      },
      Object {
        "color": "#a1a1a1",
        "name": "g",
      },
      Object {
        "color": "#b1b1b1",
        "name": "h",
      },
      Object {
        "color": "#c1c1c1",
        "name": "i",
      },
      Object {
        "color": "#d1d1d1",
        "name": "j",
      },
      Object {
        "color": "#e1e1e1",
        "name": "k",
      },
      Object {
        "color": "#f1f1f1",
        "name": "l",
      },
      Object {
        "color": "#a2a2a2",
        "name": "m",
      },
      Object {
        "color": "#b2b2b2",
        "name": "n",
      },
      Object {
        "color": "#c2c2c2",
        "name": "o",
      },
      Object {
        "color": "#d2d2d2",
        "name": "p",
      },
      Object {
        "color": "#e2e2e2",
        "name": "q",
      },
      Object {
        "color": "#f2f2f2",
        "name": "r",
      },
      Object {
        "color": "#a3a3a3",
        "name": "s",
      },
      Object {
        "color": "#b3b3b3",
        "name": "t",
      },
      Object {
        "color": "#c3c3c3",
        "name": "u",
      },
      Object {
        "color": "#d3d3d3",
        "name": "v",
      },
      Object {
        "color": "#e3e3e3",
        "name": "w",
      },
      Object {
        "color": "#f3f3f3",
        "name": "x",
      },
      Object {
        "color": "#a4a4a4",
        "name": "y",
      },
      Object {
        "color": "#b4b4b4",
        "name": "z",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build config using 'configBase' option 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build config using 'configFile' and 'configBase' options 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build config using 'configFile' option 1`] = `
Object {
  "config": Object {
    "config": Object {
      "endpoint": "http://example.com/foo/bar",
      "platform": "github",
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#111111",
        "name": "1",
      },
      Object {
        "color": "#222222",
        "name": "2",
      },
      Object {
        "color": "#333333",
        "name": "3",
      },
      Object {
        "color": "#444444",
        "name": "4",
      },
      Object {
        "color": "#555555",
        "name": "5",
      },
      Object {
        "color": "#666666",
        "name": "6",
      },
      Object {
        "color": "#777777",
        "name": "7",
      },
      Object {
        "color": "#888888",
        "name": "8",
      },
      Object {
        "color": "#999999",
        "name": "9",
      },
      Object {
        "color": "#aaaaaa",
        "name": "a",
      },
      Object {
        "color": "#bbbbbb",
        "name": "b",
      },
      Object {
        "color": "#cccccc",
        "name": "c",
      },
      Object {
        "color": "#dddddd",
        "name": "d",
      },
      Object {
        "color": "#eeeeee",
        "name": "e",
      },
      Object {
        "color": "#ffffff",
        "name": "f",
      },
      Object {
        "color": "#a1a1a1",
        "name": "g",
      },
      Object {
        "color": "#b1b1b1",
        "name": "h",
      },
      Object {
        "color": "#c1c1c1",
        "name": "i",
      },
      Object {
        "color": "#d1d1d1",
        "name": "j",
      },
      Object {
        "color": "#e1e1e1",
        "name": "k",
      },
      Object {
        "color": "#f1f1f1",
        "name": "l",
      },
      Object {
        "color": "#a2a2a2",
        "name": "m",
      },
      Object {
        "color": "#b2b2b2",
        "name": "n",
      },
      Object {
        "color": "#c2c2c2",
        "name": "o",
      },
      Object {
        "color": "#d2d2d2",
        "name": "p",
      },
      Object {
        "color": "#e2e2e2",
        "name": "q",
      },
      Object {
        "color": "#f2f2f2",
        "name": "r",
      },
      Object {
        "color": "#a3a3a3",
        "name": "s",
      },
      Object {
        "color": "#b3b3b3",
        "name": "t",
      },
      Object {
        "color": "#c3c3c3",
        "name": "u",
      },
      Object {
        "color": "#d3d3d3",
        "name": "v",
      },
      Object {
        "color": "#e3e3e3",
        "name": "w",
      },
      Object {
        "color": "#f3f3f3",
        "name": "x",
      },
      Object {
        "color": "#a4a4a4",
        "name": "y",
      },
      Object {
        "color": "#b4b4b4",
        "name": "z",
      },
    ],
  },
  "filepath": "src/lib/__tests__/fixtures/configs/simple/labelify.config.js",
}
`;

exports[`utils should success build config with array of 'extends' 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
      Object {
        "color": "#cccccc",
        "name": "2",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build config with package 'extends' 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build config with package 'extends' and 'configFile' option 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "src/lib/__tests__/fixtures/configs/package-extends/labelify.config.js",
}
`;

exports[`utils should success build with absolute path 'extends' 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build with absolute path 'extends' and 'configFile' option 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build with relative path 'extends' 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success build with relative path 'extends' and 'configFile' option 1`] = `
Object {
  "config": Object {
    "config": Object {
      "token": "IT IS INSECURITY, BETTER USE \`process.env.TOKEN\`",
    },
    "labels": Array [
      Object {
        "color": "#aaaaaa",
        "name": "0",
      },
      Object {
        "color": "#bbbbbb",
        "name": "1",
      },
    ],
  },
  "filepath": "labelify.config.js",
}
`;

exports[`utils should success get links 1`] = `Object {}`;

exports[`utils should success get links 2`] = `
Object {
  "first": "https://api.github.com/repositories/131302483/labels?page=1",
  "last": "https://api.github.com/repositories/131302483/labels?page=2",
  "next": "https://api.github.com/repositories/131302483/labels?page=2",
  "prev": "https://api.github.com/repositories/131302483/labels?page=1",
}
`;

exports[`utils should success get module path 1`] = `"node_modules/resolve-from/index.js"`;

exports[`utils should success get module path 2`] = `"node_modules/resolve-from/index.js"`;

exports[`utils should success get module path 3`] = `"Could not find \\"bar\\" extends. Do you need a \`configBasedir\`?"`;

exports[`utils should success get platform and endpoint (shortcut syntax) 1`] = `
Object {
  "endpoint": "https://api.github.com/repos/itgalaxy/labelify/labels",
  "isSelfHosted": false,
  "platform": "github",
}
`;

exports[`utils should success get platform and endpoint for custom platform and custom endpoint (2) 1`] = `
Object {
  "endpoint": "https://my.custom.platform.com/custom/api/url",
  "isSelfHosted": true,
  "platform": "gitlab",
}
`;

exports[`utils should success get platform and endpoint for custom platform and custom endpoint 1`] = `
Object {
  "endpoint": "https://my.custom.platform.com/custom/api/url",
  "isSelfHosted": true,
  "platform": "github",
}
`;

exports[`utils should success get platform and endpoint for github (git shortcut) 1`] = `
Object {
  "endpoint": "https://api.github.com/repos/111/222/labels",
  "isSelfHosted": false,
  "platform": "github",
}
`;

exports[`utils should success get platform and endpoint for github (git) 1`] = `
Object {
  "endpoint": "https://api.github.com/repos/111/222/labels",
  "isSelfHosted": false,
  "platform": "github",
}
`;

exports[`utils should success get platform and endpoint for github (self-hosted with nonstandard protocol) 1`] = `
Object {
  "endpoint": "https://mydomain.com/repos/abc/def/labels",
  "isSelfHosted": true,
  "platform": "github",
}
`;

exports[`utils should success get platform and endpoint for github (self-hosted) 1`] = `
Object {
  "endpoint": "https://github.yourcompany.com/repos/itgalaxy/labelify/labels",
  "isSelfHosted": true,
  "platform": "github",
}
`;

exports[`utils should success get platform and endpoint for github 1`] = `
Object {
  "endpoint": "https://api.github.com/repos/itgalaxy/labelify/labels",
  "isSelfHosted": false,
  "platform": "github",
}
`;

exports[`utils should success get platform and endpoint for gitlab (git shortcut) 1`] = `
Object {
  "endpoint": "https://gitlab.com/api/v4/projects/111%2F222/labels",
  "isSelfHosted": false,
  "platform": "gitlab",
}
`;

exports[`utils should success get platform and endpoint for gitlab (self-hosted with nonstandard protocol) 1`] = `
Object {
  "endpoint": "https://mydomain.com/api/v4/projects/abc%2Fdef/labels",
  "isSelfHosted": true,
  "platform": "gitlab",
}
`;

exports[`utils should success get platform and endpoint for gitlab (self-hosted) 1`] = `
Object {
  "endpoint": "https://gitlab.company.com/api/v4/projects/itgalaxy%2Flabelify/labels",
  "isSelfHosted": true,
  "platform": "gitlab",
}
`;

exports[`utils should success get platform and endpoint for gitlab 1`] = `
Object {
  "endpoint": "https://gitlab.com/api/v4/projects/itgalaxy%2Flabelify/labels",
  "isSelfHosted": false,
  "platform": "gitlab",
}
`;

exports[`utils should success get project meta (2) 1`] = `
Object {
  "endpoint": "https://api.github.com/repos/itgalaxy/labelify/labels",
  "platform": "github",
}
`;

exports[`utils should success get project meta 1`] = `
Object {
  "endpoint": "https://api.github.com/repos/itgalaxy/labelify/labels",
  "isSelfHosted": false,
  "platform": "github",
}
`;

exports[`utils should success resolve repository URL in package.json with object repository field 1`] = `"https://github.com/itgalaxy/labelify.git"`;

exports[`utils should success resolve repository URL in package.json with string repository field 1`] = `"https://github.com/itgalaxy/labelify.git"`;

exports[`utils should success walk on the links 1`] = `
Array [
  "Page 1!",
  "Page 2!",
]
`;

exports[`utils should throw error when \`package.json\` doesn't found 1`] = `"Can't find 'package.json'. Please use 'platform' and 'endpoint' options"`;

exports[`utils should throw error when config not found using \`configFile\` option 1`] = `"ENOENT: no such file or directory, open 'not-found'"`;

exports[`utils should throw error when extended config not found using \`configFile\` option 1`] = `"Could not find \\"./labelify-extended.config\\" extends. Do you need a \`configBasedir\`?"`;

exports[`utils should throw error when repository URL doesn't contains project 1`] = `"Can't resolve 'user' from 'repository' field in 'package.json'. Please correct 'repository' field."`;

exports[`utils should throw error when repository URL doesn't contains user 1`] = `"Can't resolve 'user' from 'repository' field in 'package.json'. Please correct 'repository' field."`;

exports[`utils should throw error when repository URL is not passed 1`] = `"Can't resolve your platform. Please use 'platform' options or add 'repository' field to 'package.json'"`;

exports[`utils should throw error when repository URL is not supported (2) 1`] = `"Unsupported 'bitbucket' platform"`;

exports[`utils should throw error when repository URL is not supported 1`] = `"Can't resolve your platform. Please use 'platform' options or add 'repository' field to 'package.json'"`;

exports[`utils should throw error when repository URL is self hosted platform without 'platform' option 1`] = `"Can't resolve your platform. Please use 'platform' options or add 'repository' field to 'package.json'"`;

exports[`utils should throw the error when \`package.json\` contain broken repository object field 1`] = `"Field 'repository' in 'package.json' contain invalid value"`;

exports[`utils should throw the error when \`package.json\` don't have repository field 1`] = `"Can't find 'repository' field in 'package.json'. Please use 'platform' and 'endpoint' options"`;
